qutebrowser/browser/webkit/network/pac.py

```diff
@@ -0,0 +1,529 @@
+            mask: mask for the IP address pattern informing which parts of
+                  the IP address should be matched against. 0 means ignore,
+                  255 means match.
+        """
+        host_ip = ipaddress.ip_address(host)
+        network = ipaddress.ip_network("{}/{}".format(pattern, mask))
+        return host_ip in network
+
+    @_js_slot(str)
+    def dnsResolve(self, host):
+        """Resolve a DNS hostname.
+
+        Resolves the given DNS hostname into an IP address, and returns it
+        in the dot-separated format as a string.
+
+        Args:
+            host: hostname to resolve.
+        """
+        ips = QHostInfo.fromName(host)
+        if ips.error() != QHostInfo.NoError or len(ips.addresses()) == 0:
+            return ""
+        else:
+            return ips.addresses()[0].toString()
+
+    @_js_slot()
+    def myIpAddress(self):
+        """Get host IP address.
+
+        Return the server IP address of the machine Firefox is running on,
+        as a string in the dot-separated integer format.
+        """
+        return self.dnsResolve(QHostInfo.localHostName())
+
+    @_js_slot(str)
+    def dnsDomainLevels(self, host):
+        """Count number of DNS domain levels.
+
+        Return the number (integer) of DNS domain levels (number of dots)
+        in the hostname.
+
+        Args:
+            host: the hostname from the URL.
+        """
+        return host.count('.')
+
+    @_js_slot(str, str)
+    def shExpMatch(self, mstr, shexp):
+        """Test is string matches shell expression.
+
+        Return true if the string matches the specified shell expression.
+        Currently, the patterns are shell expressions, not regular
+        expressions.
+
+        Args:
+            mstr: Any string to compare (e.g. the URL, or the hostname).
+            shexp: a shell expression to compare against.
+        """
+        return fnmatch.fnmatchcase(mstr, shexp)
+
+    def _get_dt(self, gmt):
+        if self._fixed_dt is not None:
+            if gmt:
+                return self._fixed_dt.toUTC()
+            else:
+                return self._fixed_dt
+        else:
+            if gmt:
+                return QDateTime.currentDateTimeUtc()
+            else:
+                return QDateTime.currentDateTime()
+
+    @staticmethod
+    def _in_range(avals, vals, bvals, ranges):
+        """Test if a tuple is between two tuples in numerical order."""
+        carry_a = 0
+        carry_val = 0
+        carry_b = 0
+        for a, val, b, r in zip(avals, vals, bvals, ranges):
+            if a is not None:
+                a_c = a + carry_a
+                val_c = val + carry_val
+                b_c = b + carry_b
+                if a_c < val_c < b_c:
+                    return True
+                elif not a_c <= val_c <= b_c:
+                    return False
+                carry_a += a * r
+                carry_val += val * r
+                carry_b += b * r
+        return True
```

Comment:
Couldn't this be simplified by simply using Python comparisons for the tuples? I don't get what `ranges` is exactly, but you can do `(1, 2, 3) < (4, 5, 6) < (7, 8, 9)` in Python and get `True`.